lib/controller.js

```diff
@@ -460,150 +460,121 @@ controller.connectToNetwork = function(opts) {
         });
-      }
-      // Otherwise, just resolve
-      else {
-        shaPromise = Promise.resolve();
-      }
+        // Finish up
+        return resolve();
+      } else {
 
-      // Once we have our Tessel and build version
-      return Promise.all([tesselPromise, shaPromise]).spread(function(tessel, currentVersion) {
-        // If it's not connected via USB, we can't update it
-        if (!tessel.usbConnection) {
-          return reject('Must have Tessel connected over USB to complete update. Aborting update.');
+        var version = opts.version || 'latest';
+        var tessel = Tessel.get(opts);
+        var sha = Promise.resolve();
+
+        // If we aren't forcing, we'll want to get the current SHA on Tessel
+        if (!opts.force) {
+          // Once we have the Tessel
+          // Over-ride the resolved Promise
+          sha = tessel.then(function(tessel) {
+            // Figure out what commit SHA is running on it
+            return tessel.fetchCurrentBuildInfo()
+              // Once we have the current SHA
+              .then(function(currentSHA) {
+                return updates.findBuild(builds, 'sha', currentSHA).version;
+              });
+          });
         }
 
-        // If this Tessel isn't connected to the LAN
-        if (!tessel.lanConnection || !tessel.lanConnection.authorized) {
-          // Warn the user that it's going to take a while
-          logs.warn('Warning: Authorized LAN connection not detected for this Tessel. For a faster update, cancel and connect Tessel to LAN with the `tessel wifi` command before update.');
-        }
+        // Once we have our Tessel and current version
+        return Promise.all([tessel, sha]).spread(function(tessel, currentVersion) {
+          var build = updates.findBuild(builds, 'version', version);
 
-        // If the update is forced or this build was requested
-        if (opts.build) {
-          // Fetch the requested build
-          return controller.updateTesselWithVersion(opts.force, tessel, currentVersion, requestedVersion)
-            .then(resolve, reject);
-        } else {
-          // If they have requested the latest firmware
-          var verifiedVersion;
-          if (requestedVersion === 'latest') {
-            // Figure out what the version is
-            verifiedVersion = updates.fetchBuildList()
-              .then(function(builds) {
-                // Make sure we actually have builds
-                if (Object.keys(builds).length === 0) {
-                  return reject('Could not detect any builds on server.');
-                }
-                // Sort the builds by version
-                var buildArray = updates.buildDictToSortedArray(builds);
-
-                // Return the newest one
-                return buildArray[0].version;
-              });
+          // If it's not connected via USB, we can't update it
+          if (!tessel.usbConnection) {
```

Comment:
`findBuild` doesn't return a Promise, it's operating on data that we already have from `requestBuildsList` (which does return a Promise). Either way, I think you're right about _not_ doing any of this without that connection.